rclpy/src/rclpy/_rclpy_action.c

```diff
@@ -0,0 +1,929 @@
+      (rcl_action_server_t *)PyCapsule_GetPointer(pyentity, "rcl_action_server_t");
+    ret = rcl_action_wait_set_add_action_server(wait_set, action_server, NULL);
+  } else {
+    ret = RCL_RET_ERROR;  // to avoid linter warning
+    const char * entity_name = PyCapsule_GetName(pyentity);
+    if (!entity_name) {
+      return NULL;
+    }
+    return PyErr_Format(PyExc_RuntimeError, "'%s' is not a known entity", entity_name);
+  }
+
+  if (RCL_RET_OK != ret) {
+    PyErr_Format(PyExc_RuntimeError, "Failed to add '%s' to wait set: %s",
+      PyCapsule_GetName(pyentity), rcl_get_error_string().str);
+    rcl_reset_error();
+    return NULL;
+  }
+
+  Py_RETURN_NONE;
+}
+
+/// Get the number of wait set entities that make up an action entity.
+/**
+ * \param[in] pyentity Capsule pointer to an action entity
+ *   (rcl_action_client_t or rcl_action_server_t).
+ * \return Tuple containing the number of wait set entities:
+ *   (num_subscriptions,
+ *    num_guard_conditions,
+ *    num_timers,
+ *    num_clients,
+ *    num_services)
+ */
+static PyObject *
+rclpy_action_wait_set_get_num_entities(PyObject * Py_UNUSED(self), PyObject * args)
+{
+  PyObject * pyentity;
+
+  if (!PyArg_ParseTuple(args, "O", &pyentity)) {
+    return NULL;
+  }
+
+  size_t num_subscriptions = 0u;
+  size_t num_guard_conditions = 0u;
+  size_t num_timers = 0u;
+  size_t num_clients = 0u;
+  size_t num_services = 0u;
+
+  rcl_ret_t ret;
+  if (PyCapsule_IsValid(pyentity, "rcl_action_client_t")) {
+    rcl_action_client_t * action_client =
+      (rcl_action_client_t *)PyCapsule_GetPointer(pyentity, "rcl_action_client_t");
+
+    ret = rcl_action_client_wait_set_get_num_entities(
+      action_client,
+      &num_subscriptions,
+      &num_guard_conditions,
+      &num_timers,
+      &num_clients,
+      &num_services);
+  } else if (PyCapsule_IsValid(pyentity, "rcl_action_server_t")) {
+    rcl_action_server_t * action_server =
+      (rcl_action_server_t *)PyCapsule_GetPointer(pyentity, "rcl_action_server_t");
+
+    ret = rcl_action_server_wait_set_get_num_entities(
+      action_server,
+      &num_subscriptions,
+      &num_guard_conditions,
+      &num_timers,
+      &num_clients,
+      &num_services);
+  } else {
+    ret = RCL_RET_ERROR;  // to avoid linter warning
+    const char * entity_name = PyCapsule_GetName(pyentity);
+    if (!entity_name) {
+      return NULL;
+    }
+    return PyErr_Format(PyExc_RuntimeError, "'%s' is not a known entity", entity_name);
+  }
+
+  if (RCL_RET_OK != ret) {
+    PyErr_Format(PyExc_RuntimeError,
+      "Failed to get number of entities for '%s': %s",
+      PyCapsule_GetName(pyentity),
+      rcl_get_error_string().str);
+    rcl_reset_error();
+    return NULL;
+  }
+
+  PyObject * result_tuple = PyTuple_Pack(5,
+      PyLong_FromSize_t(num_subscriptions),
+      PyLong_FromSize_t(num_guard_conditions),
+      PyLong_FromSize_t(num_timers),
+      PyLong_FromSize_t(num_clients),
+      PyLong_FromSize_t(num_services));
```

Comment:
Memory is leaked if `PyLong_FromSize_t()` succeeds but `PyTuple_Pack()` fails.

Reply:
Fixed  34db026